device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py

```diff
@@ -344,11 +393,146 @@ def reset(self, port_num):
+        self._write_byte(port_num, self.IDENTITY_EEPROM_ADDR, 0x10, 143, 0xff)
+        # Initialize datapath
+        self._write_byte(port_num, self.IDENTITY_EEPROM_ADDR, 0x10, 128, 0x00)
+        time.sleep(0.5)
+        # Validate configuration status
+        buf = self._read_eeprom_devid(port_num, self.IDENTITY_EEPROM_ADDR, self._page_to_flat(202, 0x11), 4)
+        err = "".join(buf)
+        if err != '11111111':
+            log_info("PORT {0}: ConfigErr={1}".format(port_num, err))
+            return False
+
+        return True
+
     def get_transceiver_change_event(self, timeout=0):
         """
-        TBD
+        :param timeout in milliseconds. The method is a blocking call. When timeout is 
+         zero, it only returns when there is change event, i.e., transceiver plug-in/out
+         event. When timeout is non-zero, the function can also return when the timer expires.
+         When timer expires, the return status is True and events is empty.
+        :returns: (status, events)
+        :status: Boolean, True if call successful and no system level event/error occurred, 
+         False if call not success or system level event/error occurred.
+        :events: dictionary for physical port index and the SFP status,
+         status='1' represent plug in, '0' represent plug out like {'0': '1', '31':'0'}
+         when it comes to system level event/error, the index will be '-1',
+         and status can be 'system_not_ready', 'system_become_ready', 'system_fail',
+         like {'-1':'system_not_ready'}.
         """
-        raise NotImplementedError
+        int_sfp = {}
+        end_time = time.time() + (float(timeout) / 1000.0)
+        while end_time > time.time():
+            for x in range(self.PORT_START, self.PORT_END + 1):
+                flag = self.get_presence(x)
+                if flag != self.mod_presence[x]:
+                    int_sfp[str(x)] = '1' if flag else '0'
+                    self.mod_failure[x] = 0
+                    self.mod_presence[x] = flag
+
+                # skip the following logic in case of module absent
+                if not flag:
+                    continue
+
+                # Monitoring the QSFP-DD module state, and initiate software reset when failure count > 2
+                buf = self._read_eeprom_devid(x, self.IDENTITY_EEPROM_ADDR, 0x0, 4)
+                if buf is None:
+                    continue
+                # skip, in case of QSFP28
+                if buf[0] not in "18,19":
+                    continue
+                # skip, in case that CMIS < 3.0
+                if int(buf[1], 16) < 0x30:
+                    continue
+                # advance the failure counter if state != ModuleReady
+                if ((int(buf[3], 16) >> 1) & 0x7) != 3:
+                    self.mod_failure[x] += 1
+                # initiate QSFP-DD software reset if failure counter > 2
+                if self.mod_failure[x] > 2:
```

Comment:
Should we retry inside this loop right away?  Or in the code, when is this retry called?  Another set of unplug/plug events?

Reply:
1. This is an Eoptolink specific issue, it could be stuck in ModuleLowPwr upon reinsertion, so we'll monitor its state and initiate a software reset when its state != ModuleReady for 2 runs (~3 seconds)
2. It will block the xcvrd for seconds if we retry in this loop right away, this does not seem to be a good idea.

Reply:
My question was, without later module inserts events, would this code even be called later?  I,E, the module may stuck at bad state for long time?

Reply:
The get_transceiver_change_event() is invoked by the infinite loop of the xcvrd, and no matter if a insertion/removal event generated, it will always poll the module state and trigger a software when the failure count > 2.
Note: 
1. The insertion/removal event is generated at 503, there is an in-memory flag table for the change events
2. Both the module state checker and QSFPDD config error checker are active when the modules is attached, it does not matter if the insertion/removal event generated or not.  i.e. It's outside of the if-statement at line 502